app/moderation-algo/main.py

```diff
@@ -0,0 +1,204 @@
+    # Make sure Google API is initialized
+    model = 'gemini-pro'
+
+    # First translate the question into English and Spanish. The algorithm is built to process on English questions
+    question_en, question_es, original_lang = TranslateText(question)
+
+    # Process the question through the algo and return the following fields
+    response = {
+        'question_en': question_en,
+        'question_es': question_es,
+        'original_lang': original_lang,
+        'substantive': False,
+        'offensive': False,
+        'relevant': False,
+        'topics': {},
+    }
+
+    # Check for substantiveness
+    response['substantive'] = IsSubstantive(model, question_en)
+
+    # Check for offensiveness
+    response['offensive'] = IsOffensive(question_en)
+
+    # Don't bother with further processing if the question is unsubstantive or offensive
+    if(response['offensive'] or not response['substantive']):
+        return response
+
+    # Check for relevancy if an issue was provided
+    if(issue):
+        response['relevant'] = IsRelevant(model, question_en, issue)
+
+    # Classify into topics if any were provided
+    if(topics):
+        for topic, description in topics.items():
+            response['topics'][topic] = DoesQuestionFitCategory(model, question_en, topic, description)
+
+    return response
+
+def ConnectToRedis() -> redis.Redis:
+    "Establish a connection to Redis for temporary persistent memory. Returns the connection object."
+    REDIS_HOST='localhost'
+    REDIS_PORT=6379
+    REDIS_USERNAME='default'
+    REDIS_PASSWORD='redispassword'
+    return redis.StrictRedis(host=REDIS_HOST, port=REDIS_PORT, username=REDIS_USERNAME, password=REDIS_PASSWORD,
+                             charset='utf-8', decode_responses=True)
+
+@app.route('/', methods=['POST'])
+def HandleUserInput():
+    # Check if the request contains JSON data
+    if request.is_json:
+        # NOTE: Make sure Google API is initialized at this point
+        model = 'gemini-pro'
+
+        # Connect to Redis with a week expiration time
+        secInAWeek = 604800
+        r = ConnectToRedis()
+
+        # Decode which stage the request is going to
+        stage = request.get_json().get('stage')
+
+        # Stage 1: Extract the overall issue and the topics and definitions from reading materials
+        if(stage == 'extraction'):
+            # Extract and issue and topics for later use
+            reading_materials = request.get_json().get('reading_materials')
+            if(not reading_materials):
+                return jsonify({'ERROR': 'Missing field "reading_materials" in request data'}), 422 # HTTP unprocessable Entity
+            issue = ExtractIssueFromReadingMaterials(model, reading_materials)
+            topics = ExtractTopicsDescriptions(model, reading_materials)
+
+            # Save them for later use, expiring in a week
+            r.set('moderation_issue', issue, ex=secInAWeek)
```

Comment:
All the redis keys should be unique to the event, can use something like `moderation_issue_${eventId}`.